File: collector/test/action-window/reply-submission/review-body-fingerprint.test.ts
/**
 * `review-body-fingerprint/v1` — the collector side of the shared golden-vector proof. Loading the SAME
 * `contracts/review-fingerprint/v1/golden-vectors.json` that the Java `ReviewBodyFingerprintTest` loads, and
 * both passing, is the cross-language equivalence proof.
 */
import { describe, it, expect } from "vitest";
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, resolve } from "node:path";
import {
  normalizeForFingerprint,
  reviewBodyFingerprint,
} from "../../../src/action-window/reply-submission/review-body-fingerprint";

const HERE = dirname(fileURLToPath(import.meta.url));
interface Vector { name: string; raw: string; normalized: string; fingerprint: string }
const VECTORS = JSON.parse(
  readFileSync(resolve(HERE, "../../../../contracts/review-fingerprint/v1/golden-vectors.json"), "utf8"),
) as { spec: string; cases: Vector[] };

describe("review-body-fingerprint/v1 — shared synthetic golden vectors", () => {
  it("loads the v1 spec vectors", () => {
    expect(VECTORS.spec).toBe("review-body-fingerprint/v1");
    expect(VECTORS.cases.length).toBeGreaterThanOrEqual(8);
  });

  for (const c of VECTORS.cases) {
    it(`normalizes '${c.name}' to the golden form`, () => {
      expect(normalizeForFingerprint(c.raw)).toBe(c.normalized);
    });
    it(`fingerprints '${c.name}' to the golden hash`, () => {
      expect(c.fingerprint).toMatch(/^[0-9a-f]{64}$/);
      expect(reviewBodyFingerprint(c.raw)).toBe(c.fingerprint);
    });
  }

  it("CRLF and LF bodies fingerprint identically", () => {
    const crlf = VECTORS.cases.find((c) => c.name === "crlf")!;
    const lf = VECTORS.cases.find((c) => c.name === "lf-equiv-crlf")!;
    expect(reviewBodyFingerprint(crlf.raw)).toBe(reviewBodyFingerprint(lf.raw));
  });

  it("the PII case tokenizes volatile spans and no raw span survives normalization", () => {
    const pii = VECTORS.cases.find((c) => c.name === "pii-body")!;
    const norm = normalizeForFingerprint(pii.raw);
    for (const tok of ["[링크]", "[이메일]", "[전화번호]", "[번호]"]) expect(norm).toContain(tok);
    for (const raw of ["naver.me", "hong@test.com", "010-1234-5678", "02-345-6789", "1234567890"]) {
      expect(norm).not.toContain(raw);
    }
  });

  it("empty/whitespace-only input yields a valid 64-hex fingerprint", () => {
    expect(reviewBodyFingerprint("")).toMatch(/^[0-9a-f]{64}$/);
    expect(normalizeForFingerprint("   \n\t  ")).toBe("");
  });
});

File: backend/src/test/java/com/sellerops/common/ReviewBodyFingerprintTest.java
package com.sellerops.common;

import static org.assertj.core.api.Assertions.assertThat;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import org.junit.jupiter.api.DynamicTest;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestFactory;

/**
 * The backend side of the {@code review-body-fingerprint/v1} cross-language proof. It loads the SAME
 * {@code contracts/review-fingerprint/v1/golden-vectors.json} that the collector's
 * {@code review-body-fingerprint.test.ts} loads; both passing is the equivalence guarantee. All vectors are
 * SYNTHETIC — no real PII.
 */
class ReviewBodyFingerprintTest {

    private static final ObjectMapper MAPPER = new ObjectMapper();

    private static Path locateVectors() {
        Path dir = Paths.get(System.getProperty("user.dir")).toAbsolutePath();
        for (int i = 0; i < 8 && dir != null; i++, dir = dir.getParent()) {
            Path candidate = dir.resolve("contracts/review-fingerprint/v1/golden-vectors.json");
            if (Files.exists(candidate)) {
                return candidate;
            }
        }
        throw new IllegalStateException(
                "golden-vectors.json not found from user.dir=" + System.getProperty("user.dir"));
    }

    private static JsonNode cases() throws Exception {
        JsonNode root = MAPPER.readTree(Files.readString(locateVectors()));
        assertThat(root.get("spec").asText()).isEqualTo("review-body-fingerprint/v1");
        return root.get("cases");
    }

    @TestFactory
    List<DynamicTest> reproducesEverySharedGoldenVector() throws Exception {
        List<DynamicTest> tests = new ArrayList<>();
        for (JsonNode c : cases()) {
            String name = c.get("name").asText();
            String raw = c.get("raw").asText();
            String normalized = c.get("normalized").asText();
            String fingerprint = c.get("fingerprint").asText();
            tests.add(DynamicTest.dynamicTest("normalize:" + name, () ->
                    assertThat(ReviewBodyFingerprint.normalizeForFingerprint(raw)).isEqualTo(normalized)));
            tests.add(DynamicTest.dynamicTest("fingerprint:" + name, () -> {
                assertThat(fingerprint).matches("[0-9a-f]{64}");
                assertThat(ReviewBodyFingerprint.of(raw)).isEqualTo(fingerprint);
            }));
        }
        assertThat(tests).hasSizeGreaterThanOrEqualTo(16);
        return tests;
    }

    @Test
    void piiCaseTokenizesVolatileSpansAndLeaksNoRawSpan() throws Exception {
        JsonNode pii = null;
        for (JsonNode c : cases()) {
            if (c.get("name").asText().equals("pii-body")) {
                pii = c;
            }
        }
        assertThat(pii).isNotNull();
        String norm = ReviewBodyFingerprint.normalizeForFingerprint(pii.get("raw").asText());
        assertThat(norm)
                .contains("[링크]", "[이메일]", "[전화번호]", "[번호]")
                .doesNotContain("naver.me", "hong@test.com", "010-1234-5678", "02-345-6789", "1234567890");
    }

    @Test
    void nullAndBlankAreSafe() {
        assertThat(ReviewBodyFingerprint.normalizeForFingerprint(null)).isEmpty();
        assertThat(ReviewBodyFingerprint.of("")).matches("[0-9a-f]{64}");
    }
}
